src/NServiceBus.Transport.SQS/InputQueuePump.cs

```diff
@@ -308,69 +317,111 @@ internal async Task ProcessMessage(Message receivedMessage, CancellationToken me
+                    if (errorHandlerResult == ErrorHandleResult.RetryRequired)
+                    {
+                        await ReturnMessageToQueue(nativeMessage).ConfigureAwait(false);
 
-                    messageProcessedOk = true;
+                        return;
+                    }
                 }
-                catch (Exception ex) when (!ex.IsCausedBy(messageProcessingCancellationToken))
+                catch (Exception onErrorEx) when (!onErrorEx.IsCausedBy(messageProcessingCancellationToken))
                 {
-                    immediateProcessingAttempts++;
-                    var errorHandlerResult = ErrorHandleResult.RetryRequired;
+                    criticalErrorAction(
+                        $"Failed to execute recoverability policy for message with native ID: `{nativeMessageId}`",
+                        onErrorEx,
+                        messageProcessingCancellationToken);
 
-                    try
-                    {
-                        errorHandlerResult = await onError(new ErrorContext(ex,
-                            new Dictionary<string, string>(headers),
-                            nativeMessageId,
-                            body,
-                            transportTransaction,
-                            immediateProcessingAttempts,
-                            ReceiveAddress,
-                            context), messageProcessingCancellationToken).ConfigureAwait(false);
-                    }
-                    catch (Exception onErrorEx) when (!onErrorEx.IsCausedBy(messageProcessingCancellationToken))
-                    {
-                        criticalErrorAction($"Failed to execute recoverability policy for message with native ID: `{nativeMessageId}`", onErrorEx, messageProcessingCancellationToken);
-                    }
+                    await ReturnMessageToQueue(nativeMessage).ConfigureAwait(false);
 
-                    errorHandled = errorHandlerResult == ErrorHandleResult.Handled;
+                    return;
                 }
             }
+
+            await DeleteMessage(nativeMessage, s3BodyKey, messageProcessingCancellationToken).ConfigureAwait(false);
+        }
+
+
+#pragma warning disable PS0018 //Cancellation token intentionally not passed because cancellation shouldn't stop messages from being returned to the queue
+        async Task ReturnMessageToQueue(Message message)
+#pragma warning restore PS0018
+        {
+            try
+            {
+                await sqsClient.ChangeMessageVisibilityAsync(
+                    new ChangeMessageVisibilityRequest()
+                    {
+                        QueueUrl = inputQueueUrl,
+                        ReceiptHandle = message.ReceiptHandle,
+                        VisibilityTimeout = 0
+                    }).ConfigureAwait(false);
+            }
+            catch (ReceiptHandleIsInvalidException ex)
```

Comment:
Not sure these specialized `catch` blocks are really adding anything here. Let's simplify this down to just the `Exception` one.